File: app/src/main/java/info/nightscout/androidaps/plugins/pump/common/PumpPluginAbstract.java
package info.nightscout.androidaps.plugins.pump.common;

import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;

import androidx.fragment.app.FragmentActivity;

import org.json.JSONException;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Date;

import info.nightscout.androidaps.BuildConfig;
import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.data.DetailedBolusInfo;
import info.nightscout.androidaps.data.Profile;
import info.nightscout.androidaps.data.PumpEnactResult;
import info.nightscout.androidaps.db.ExtendedBolus;
import info.nightscout.androidaps.db.TemporaryBasal;
import info.nightscout.androidaps.events.EventAppExit;
import info.nightscout.androidaps.interfaces.ConstraintsInterface;
import info.nightscout.androidaps.interfaces.PluginBase;
import info.nightscout.androidaps.interfaces.PluginDescription;
import info.nightscout.androidaps.interfaces.PluginType;
import info.nightscout.androidaps.interfaces.PumpDescription;
import info.nightscout.androidaps.interfaces.PumpInterface;
import info.nightscout.androidaps.logging.L;
import info.nightscout.androidaps.plugins.bus.RxBus;
import info.nightscout.androidaps.plugins.general.overview.events.EventOverviewBolusProgress;
import info.nightscout.androidaps.plugins.pump.common.data.PumpStatus;
import info.nightscout.androidaps.plugins.pump.common.defs.PumpDriverState;
import info.nightscout.androidaps.plugins.pump.common.defs.PumpType;
import info.nightscout.androidaps.plugins.pump.medtronic.data.MedtronicHistoryData;
import info.nightscout.androidaps.plugins.treatments.Treatment;
import info.nightscout.androidaps.plugins.treatments.TreatmentsPlugin;
import info.nightscout.androidaps.utils.DateUtil;
import info.nightscout.androidaps.utils.DecimalFormatter;
import info.nightscout.androidaps.utils.FabricPrivacy;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by andy on 23.04.18.
 */

// When using this class, make sure that your first step is to create mConnection (see MedtronicPumpPlugin)

public abstract class PumpPluginAbstract extends PluginBase implements PumpInterface, ConstraintsInterface {
    private CompositeDisposable disposable = new CompositeDisposable();

    private static final Logger LOG = LoggerFactory.getLogger(L.PUMP);

    protected static final PumpEnactResult OPERATION_NOT_SUPPORTED = new PumpEnactResult().success(false)
            .enacted(false).comment(MainApp.gs(R.string.pump_operation_not_supported_by_pump_driver));
    protected static final PumpEnactResult OPERATION_NOT_YET_SUPPORTED = new PumpEnactResult().success(false)
            .enacted(false).comment(MainApp.gs(R.string.pump_operation_not_yet_supported_by_pump));

    protected PumpDescription pumpDescription = new PumpDescription();
    protected PumpStatus pumpStatus;
    protected ServiceConnection serviceConnection = null;
    protected boolean serviceRunning = false;
    // protected boolean isInitialized = false;
    protected PumpDriverState pumpState = PumpDriverState.NotInitialized;
    protected boolean displayConnectionMessages = false;


    protected PumpPluginAbstract(PluginDescription pluginDescription, PumpType pumpType) {

        super(pluginDescription);

        pumpDescription.setPumpDescription(pumpType);

        initPumpStatusData();

    }


    public abstract void initPumpStatusData();


    @Override
    protected void onStart() {
        super.onStart();
        Context context = MainApp.instance().getApplicationContext();
        Intent intent = new Intent(context, getServiceClass());
        context.bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);

        serviceRunning = true;

        disposable.add(RxBus.INSTANCE
                .toObservable(EventAppExit.class)
                .observeOn(Schedulers.io())
                .subscribe(event -> {
                    MainApp.instance().getApplicationContext().unbindService(serviceConnection);
                }, FabricPrivacy::logException)
        );
        onStartCustomActions();
    }


    @Override
    protected void onStop() {
        Context context = MainApp.instance().getApplicationContext();
        context.unbindService(serviceConnection);

        serviceRunning = false;

        disposable.clear();
        super.onStop();
    }


    /**
     * If we need to run any custom actions in onStart (triggering events, etc)
     */
    public abstract void onStartCustomActions();

    @Override
    public void switchAllowed(boolean newState, FragmentActivity activity, PluginType type) {
        confirmPumpPluginActivation(newState, activity, type);
    }

    /**
     * Service class (same one you did serviceConnection for)
     *
     * @return
     */
    public abstract Class getServiceClass();

    public PumpStatus getPumpStatusData() {
        return pumpStatus;
    }


    public boolean isInitialized() {
        return PumpDriverState.isInitialized(pumpState);
    }


    public boolean isSuspended() {
        return pumpState == PumpDriverState.Suspended;
    }


    public boolean isBusy() {
        return pumpState == PumpDriverState.Busy;
    }


    public boolean isConnected() {
        if (displayConnectionMessages && isLoggingEnabled())
            LOG.warn("isConnected [PumpPluginAbstract].");
        return PumpDriverState.isConnected(pumpState);
    }


    public boolean isConnecting() {
        if (displayConnectionMessages && isLoggingEnabled())
            LOG.warn("isConnecting [PumpPluginAbstract].");
        return pumpState == PumpDriverState.Connecting;
    }


    public void connect(String reason) {
        if (displayConnectionMessages && isLoggingEnabled())
            LOG.warn("connect (reason={}) [PumpPluginAbstract] - default (empty) implementation.", reason);
    }


    public void disconnect(String reason) {
        if (displayConnectionMessages && isLoggingEnabled())
            LOG.warn("disconnect (reason={}) [PumpPluginAbstract] - default (empty) implementation.", reason);
    }


    public void stopConnecting() {
        if (displayConnectionMessages && isLoggingEnabled())
            LOG.warn("stopConnecting [PumpPluginAbstract] - default (empty) implementation.");
    }


    @Override
    public boolean isHandshakeInProgress() {
        if (displayConnectionMessages && isLoggingEnabled())
            LOG.warn("isHandshakeInProgress [PumpPluginAbstract] - default (empty) implementation.");
        return false;
    }


    @Override
    public void finishHandshaking() {
        if (displayConnectionMessages && isLoggingEnabled())
            LOG.warn("finishHandshaking [PumpPluginAbstract] - default (empty) implementation.");
    }


    public void getPumpStatus() {
        if (isLoggingEnabled())
            LOG.warn("getPumpStatus [PumpPluginAbstract] - Not implemented.");
    }


    // Upload to pump new basal profile
    public PumpEnactResult setNewBasalProfile(Profile profile) {
        if (isLoggingEnabled())
            LOG.warn("setNewBasalProfile [PumpPluginAbstract] - Not implemented.");
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver);
    }


    public boolean isThisProfileSet(Profile profile) {
        if (isLoggingEnabled())
            LOG.warn("isThisProfileSet [PumpPluginAbstract] - Not implemented.");
        return true;
    }


    public long lastDataTime() {
        if (isLoggingEnabled())
            LOG.warn("lastDataTime [PumpPluginAbstract].");
        return pumpStatus.lastConnection;
    }


    public double getBaseBasalRate() {
        if (isLoggingEnabled())
            LOG.warn("getBaseBasalRate [PumpPluginAbstract] - Not implemented.");
        return 0.0d;
    } // base basal rate, not temp basal


    public void stopBolusDelivering() {
        if (isLoggingEnabled())
            LOG.warn("stopBolusDelivering [PumpPluginAbstract] - Not implemented.");
    }


    @Override
    public PumpEnactResult setTempBasalAbsolute(Double absoluteRate, Integer durationInMinutes, Profile profile,
                                                boolean enforceNew) {
        if (isLoggingEnabled())
            LOG.warn("setTempBasalAbsolute [PumpPluginAbstract] - Not implemented.");
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver);
    }


    @Override
    public PumpEnactResult setTempBasalPercent(Integer percent, Integer durationInMinutes, Profile profile,
                                               boolean enforceNew) {
        if (isLoggingEnabled())
            LOG.warn("setTempBasalPercent [PumpPluginAbstract] - Not implemented.");
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver);
    }


    public PumpEnactResult setExtendedBolus(Double insulin, Integer durationInMinutes) {
        if (isLoggingEnabled())
            LOG.warn("setExtendedBolus [PumpPluginAbstract] - Not implemented.");
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver);
    }


    // some pumps might set a very short temp close to 100% as cancelling a temp can be noisy
    // when the cancel request is requested by the user (forced), the pump should always do a real cancel

    public PumpEnactResult cancelTempBasal(boolean enforceNew) {
        if (isLoggingEnabled())
            LOG.warn("cancelTempBasal [PumpPluginAbstract] - Not implemented.");
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver);
    }


    public PumpEnactResult cancelExtendedBolus() {
        if (isLoggingEnabled())
            LOG.warn("cancelExtendedBolus [PumpPluginAbstract] - Not implemented.");
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver);
    }


    // Status to be passed to NS

    // public JSONObject getJSONStatus(Profile profile, String profileName) {
    // return pumpDriver.getJSONStatus(profile, profileName);
    // }

    public String deviceID() {
        if (isLoggingEnabled())
            LOG.warn("deviceID [PumpPluginAbstract] - Not implemented.");
        return "FakeDevice";
    }


    // Pump capabilities

    public PumpDescription getPumpDescription() {
        return pumpDescription;
    }


    // Short info for SMS, Wear etc

    public boolean isFakingTempsByExtendedBoluses() {
        if (isLoggingEnabled())
            LOG.warn("isFakingTempsByExtendedBoluses [PumpPluginAbstract] - Not implemented.");
        return false;
    }


    @Override
    public PumpEnactResult loadTDDs() {
        if (isLoggingEnabled())
            LOG.warn("loadTDDs [PumpPluginAbstract] - Not implemented.");
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver);
    }


    @Override
    public JSONObject getJSONStatus(Profile profile, String profileName) {

        long now = System.currentTimeMillis();
        if ((pumpStatus.lastConnection + 5 * 60 * 1000L) < System.currentTimeMillis()) {
            return null;
        }

        JSONObject pump = new JSONObject();
        JSONObject battery = new JSONObject();
        JSONObject status = new JSONObject();
        JSONObject extended = new JSONObject();
        try {
            battery.put("percent", pumpStatus.batteryRemaining);
            status.put("status", pumpStatus.pumpStatusType != null ? pumpStatus.pumpStatusType.getStatus() : "normal");
            extended.put("Version", BuildConfig.VERSION_NAME + "-" + BuildConfig.BUILDVERSION);
            try {
                extended.put("ActiveProfile", profileName);
            } catch (Exception e) {
            }

            TemporaryBasal tb = TreatmentsPlugin.getPlugin().getTempBasalFromHistory(System.currentTimeMillis());
            if (tb != null) {
                extended.put("TempBasalAbsoluteRate",
                        tb.tempBasalConvertedToAbsolute(System.currentTimeMillis(), profile));
                extended.put("TempBasalStart", DateUtil.dateAndTimeString(tb.date));
                extended.put("TempBasalRemaining", tb.getPlannedRemainingMinutes());
            }

            ExtendedBolus eb = TreatmentsPlugin.getPlugin().getExtendedBolusFromHistory(System.currentTimeMillis());
            if (eb != null) {
                extended.put("ExtendedBolusAbsoluteRate", eb.absoluteRate());
                extended.put("ExtendedBolusStart", DateUtil.dateAndTimeString(eb.date));
                extended.put("ExtendedBolusRemaining", eb.getPlannedRemainingMinutes());
            }

            status.put("timestamp", DateUtil.toISOString(new Date()));

            pump.put("battery", battery);
            pump.put("status", status);
            pump.put("extended", extended);
            pump.put("reservoir", pumpStatus.reservoirRemainingUnits);
            pump.put("clock", DateUtil.toISOString(new Date()));
        } catch (JSONException e) {
            LOG.error("Unhandled exception", e);
        }
        return pump;
    }


    // FIXME i18n, null checks: iob, TDD
    @Override
    public String shortStatus(boolean veryShort) {
        String ret = "";
        if (pumpStatus.lastConnection != 0) {
            Long agoMsec = System.currentTimeMillis() - pumpStatus.lastConnection;
            int agoMin = (int) (agoMsec / 60d / 1000d);
            ret += "LastConn: " + agoMin + " min ago\n";
        }
        if (pumpStatus.lastBolusTime != null && pumpStatus.lastBolusTime.getTime() != 0) {
            ret += "LastBolus: " + DecimalFormatter.to2Decimal(pumpStatus.lastBolusAmount) + "U @" + //
                    android.text.format.DateFormat.format("HH:mm", pumpStatus.lastBolusTime) + "\n";
        }
        TemporaryBasal activeTemp = TreatmentsPlugin.getPlugin()
                .getRealTempBasalFromHistory(System.currentTimeMillis());
        if (activeTemp != null) {
            ret += "Temp: " + activeTemp.toStringFull() + "\n";
        }
        ExtendedBolus activeExtendedBolus = TreatmentsPlugin.getPlugin().getExtendedBolusFromHistory(
                System.currentTimeMillis());
        if (activeExtendedBolus != null) {
            ret += "Extended: " + activeExtendedBolus.toString() + "\n";
        }
        // if (!veryShort) {
        // ret += "TDD: " + DecimalFormatter.to0Decimal(pumpStatus.dailyTotalUnits) + " / "
        // + pumpStatus.maxDailyTotalUnits + " U\n";
        // }
        ret += "IOB: " + pumpStatus.iob + "U\n";
        ret += "Reserv: " + DecimalFormatter.to0Decimal(pumpStatus.reservoirRemainingUnits) + "U\n";
        ret += "Batt: " + pumpStatus.batteryRemaining + "\n";
        return ret;
    }


    @Override
    public PumpEnactResult deliverTreatment(DetailedBolusInfo detailedBolusInfo) {

        try {
            if (detailedBolusInfo.insulin == 0 && detailedBolusInfo.carbs == 0) {
                // neither carbs nor bolus requested
                if (isLoggingEnabled())
                    LOG.error("deliverTreatment: Invalid input");
                return new PumpEnactResult().success(false).enacted(false).bolusDelivered(0d).carbsDelivered(0d)
                        .comment(MainApp.gs(R.string.danar_invalidinput));
            } else if (detailedBolusInfo.insulin > 0) {
                // bolus needed, ask pump to deliver it
                return deliverBolus(detailedBolusInfo);
            } else {
                if (MedtronicHistoryData.doubleBolusDebug)
                    LOG.debug("DoubleBolusDebug: deliverTreatment::(carb only entry)");

                // no bolus required, carb only treatment
                TreatmentsPlugin.getPlugin().addToHistoryTreatment(detailedBolusInfo, true);

                EventOverviewBolusProgress bolusingEvent = EventOverviewBolusProgress.INSTANCE;
                bolusingEvent.setT(new Treatment());
                bolusingEvent.getT().isSMB = detailedBolusInfo.isSMB;
                bolusingEvent.setPercent(100);
                RxBus.INSTANCE.send(bolusingEvent);

                if (isLoggingEnabled())
                    LOG.debug("deliverTreatment: Carb only treatment.");

                return new PumpEnactResult().success(true).enacted(true).bolusDelivered(0d)
                        .carbsDelivered(detailedBolusInfo.carbs).comment(MainApp.gs(R.string.virtualpump_resultok));
            }
        } finally {
            triggerUIChange();
        }

    }


    public boolean isLoggingEnabled() {
        return L.isEnabled(L.PUMP);
    }


    protected abstract PumpEnactResult deliverBolus(DetailedBolusInfo detailedBolusInfo);


    protected abstract void triggerUIChange();


    public static PumpEnactResult getOperationNotSupportedWithCustomText(int resourceId) {
        return new PumpEnactResult().success(false).enacted(false).comment(MainApp.gs(resourceId));
    }

}

File: pump/pump-common/src/main/java/info/nightscout/pump/common/PumpPluginAbstract.kt
package info.nightscout.pump.common

import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.text.format.DateFormat
import com.google.gson.GsonBuilder
import dagger.android.HasAndroidInjector
import info.nightscout.core.utils.fabric.FabricPrivacy
import info.nightscout.interfaces.constraints.Constraints
import info.nightscout.interfaces.plugin.ActivePlugin
import info.nightscout.interfaces.plugin.PluginDescription
import info.nightscout.interfaces.profile.Profile
import info.nightscout.interfaces.pump.DetailedBolusInfo
import info.nightscout.interfaces.pump.Pump
import info.nightscout.interfaces.pump.PumpEnactResult
import info.nightscout.interfaces.pump.PumpPluginBase
import info.nightscout.interfaces.pump.PumpSync
import info.nightscout.interfaces.pump.PumpSync.TemporaryBasalType
import info.nightscout.interfaces.pump.defs.ManufacturerType
import info.nightscout.interfaces.pump.defs.PumpDescription
import info.nightscout.interfaces.pump.defs.PumpType
import info.nightscout.interfaces.queue.CommandQueue
import info.nightscout.interfaces.utils.DecimalFormatter.to0Decimal
import info.nightscout.interfaces.utils.DecimalFormatter.to2Decimal
import info.nightscout.pump.common.data.PumpStatus
import info.nightscout.pump.common.defs.PumpDriverState
import info.nightscout.pump.common.sync.PumpDbEntryCarbs
import info.nightscout.pump.common.sync.PumpSyncStorage
import info.nightscout.rx.AapsSchedulers
import info.nightscout.rx.bus.RxBus
import info.nightscout.rx.events.EventAppExit
import info.nightscout.rx.events.EventCustomActionsChanged
import info.nightscout.rx.events.EventOverviewBolusProgress
import info.nightscout.rx.logging.AAPSLogger
import info.nightscout.rx.logging.LTag
import info.nightscout.shared.interfaces.ResourceHelper
import info.nightscout.shared.sharedPreferences.SP
import info.nightscout.shared.utils.DateUtil
import io.reactivex.rxjava3.disposables.CompositeDisposable
import org.json.JSONException
import org.json.JSONObject

/**
 * Created by andy on 23.04.18.
 */
// When using this class, make sure that your first step is to create mConnection (see MedtronicPumpPlugin)
abstract class PumpPluginAbstract protected constructor(
    pluginDescription: PluginDescription,
    pumpType: PumpType,
    injector: HasAndroidInjector,
    rh: ResourceHelper,
    aapsLogger: AAPSLogger,
    commandQueue: CommandQueue,
    var rxBus: RxBus,
    var activePlugin: ActivePlugin,
    var sp: SP,
    var context: Context,
    var fabricPrivacy: FabricPrivacy,
    var dateUtil: DateUtil,
    var aapsSchedulers: AapsSchedulers,
    var pumpSync: PumpSync,
    var pumpSyncStorage: PumpSyncStorage
) : PumpPluginBase(pluginDescription, injector, aapsLogger, rh, commandQueue), Pump, Constraints, info.nightscout.pump.common.sync.PumpSyncEntriesCreator {

    protected val disposable = CompositeDisposable()

    // Pump capabilities
    final override var pumpDescription = PumpDescription()
    //protected set

    protected var serviceConnection: ServiceConnection? = null
    protected var serviceRunning = false
    protected var pumpState = PumpDriverState.NotInitialized
    protected var displayConnectionMessages = false

    var pumpType: PumpType = PumpType.GENERIC_AAPS
        set(value) {
            field = value
            pumpDescription.fillFor(value)
        }

    protected var gson = GsonBuilder().excludeFieldsWithoutExposeAnnotation().create()

    abstract fun initPumpStatusData()

    open fun hasService(): Boolean {
        return true
    }

    override fun onStart() {
        super.onStart()
        initPumpStatusData()
        if (hasService()) {
            val intent = Intent(context, serviceClass)
            context.bindService(intent, serviceConnection!!, Context.BIND_AUTO_CREATE)
            disposable.add(rxBus
                               .toObservable(EventAppExit::class.java)
                               .observeOn(aapsSchedulers.io)
                               .subscribe({ _ -> context.unbindService(serviceConnection!!) }) { throwable: Throwable? -> fabricPrivacy.logException(throwable!!) }
            )
        }
        serviceRunning = true
        onStartScheduledPumpActions()
    }

    override fun onStop() {
        aapsLogger.debug(LTag.PUMP, model().model + " onStop()")
        if (hasService()) {
            context.unbindService(serviceConnection!!)
        }
        serviceRunning = false
        disposable.clear()
        super.onStop()
    }

    /**
     * If we need to run any custom actions in onStart (triggering events, etc)
     */
    abstract fun onStartScheduledPumpActions()

    /**
     * Service class (same one you did serviceConnection for)
     *
     * @return Class
     */
    abstract val serviceClass: Class<*>?
    abstract val pumpStatusData: PumpStatus

    override fun isInitialized(): Boolean = pumpState.isInitialized()
    override fun isSuspended(): Boolean = pumpState == PumpDriverState.Suspended
    override fun isBusy(): Boolean = pumpState == PumpDriverState.Busy

    override fun isConnected(): Boolean {
        if (displayConnectionMessages) aapsLogger.debug(LTag.PUMP, "isConnected [PumpPluginAbstract].")
        return pumpState.isConnected()
    }

    override fun isConnecting(): Boolean {
        if (displayConnectionMessages) aapsLogger.debug(LTag.PUMP, "isConnecting [PumpPluginAbstract].")
        return pumpState === PumpDriverState.Connecting
    }

    override fun connect(reason: String) {
        if (displayConnectionMessages) aapsLogger.debug(LTag.PUMP, "connect (reason={}) [PumpPluginAbstract] - default (empty) implementation.$reason")
    }

    override fun disconnect(reason: String) {
        if (displayConnectionMessages) aapsLogger.debug(LTag.PUMP, "disconnect (reason={}) [PumpPluginAbstract] - default (empty) implementation.$reason")
    }

    override fun stopConnecting() {
        if (displayConnectionMessages) aapsLogger.debug(LTag.PUMP, "stopConnecting [PumpPluginAbstract] - default (empty) implementation.")
    }

    override fun isHandshakeInProgress(): Boolean {
        if (displayConnectionMessages) aapsLogger.debug(LTag.PUMP, "isHandshakeInProgress [PumpPluginAbstract] - default (empty) implementation.")
        return false
    }

    override fun finishHandshaking() {
        if (displayConnectionMessages) aapsLogger.debug(LTag.PUMP, "finishHandshaking [PumpPluginAbstract] - default (empty) implementation.")
    }

    // Upload to pump new basal profile
    override fun setNewBasalProfile(profile: Profile): PumpEnactResult {
        aapsLogger.debug(LTag.PUMP, "setNewBasalProfile [PumpPluginAbstract] - Not implemented.")
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver)
    }

    override fun isThisProfileSet(profile: Profile): Boolean {
        aapsLogger.debug(LTag.PUMP, "isThisProfileSet [PumpPluginAbstract] - Not implemented.")
        return true
    }

    override fun lastDataTime(): Long {
        aapsLogger.debug(LTag.PUMP, "lastDataTime [PumpPluginAbstract].")
        return pumpStatusData.lastConnection
    }

    // base basal rate, not temp basal
    override val baseBasalRate: Double
        get() {
            aapsLogger.debug(LTag.PUMP, "getBaseBasalRate [PumpPluginAbstract] - Not implemented.")
            return 0.0
        }

    override fun stopBolusDelivering() {
        aapsLogger.debug(LTag.PUMP, "stopBolusDelivering [PumpPluginAbstract] - Not implemented.")
    }

    override fun setTempBasalAbsolute(absoluteRate: Double, durationInMinutes: Int, profile: Profile, enforceNew: Boolean, tbrType: TemporaryBasalType): PumpEnactResult {
        aapsLogger.debug(LTag.PUMP, "setTempBasalAbsolute [PumpPluginAbstract] - Not implemented.")
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver)
    }

    override fun setTempBasalPercent(percent: Int, durationInMinutes: Int, profile: Profile, enforceNew: Boolean, tbrType: TemporaryBasalType): PumpEnactResult {
        aapsLogger.debug(LTag.PUMP, "setTempBasalPercent [PumpPluginAbstract] - Not implemented.")
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver)
    }

    override fun setExtendedBolus(insulin: Double, durationInMinutes: Int): PumpEnactResult {
        aapsLogger.debug(LTag.PUMP, "setExtendedBolus [PumpPluginAbstract] - Not implemented.")
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver)
    }

    // some pumps might set a very short temp close to 100% as cancelling a temp can be noisy
    // when the cancel request is requested by the user (forced), the pump should always do a real cancel
    override fun cancelTempBasal(enforceNew: Boolean): PumpEnactResult {
        aapsLogger.debug(LTag.PUMP, "cancelTempBasal [PumpPluginAbstract] - Not implemented.")
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver)
    }

    override fun cancelExtendedBolus(): PumpEnactResult {
        aapsLogger.debug(LTag.PUMP, "cancelExtendedBolus [PumpPluginAbstract] - Not implemented.")
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver)
    }

    // Status to be passed to NS
    // public JSONObject getJSONStatus(Profile profile, String profileName) {
    // return pumpDriver.getJSONStatus(profile, profileName);
    // }
    open fun deviceID(): String {
        aapsLogger.debug(LTag.PUMP, "deviceID [PumpPluginAbstract] - Not implemented.")
        return "FakeDevice"
    }

    // Short info for SMS, Wear etc
    override val isFakingTempsByExtendedBoluses: Boolean
        get() {
            aapsLogger.debug(LTag.PUMP, "isFakingTempsByExtendedBoluses [PumpPluginAbstract] - Not implemented.")
            return false
        }

    override fun loadTDDs(): PumpEnactResult {
        aapsLogger.debug(LTag.PUMP, "loadTDDs [PumpPluginAbstract] - Not implemented.")
        return getOperationNotSupportedWithCustomText(R.string.pump_operation_not_supported_by_pump_driver)
    }

    override fun getJSONStatus(profile: Profile, profileName: String, version: String): JSONObject {
        if (pumpStatusData.lastConnection + 60 * 60 * 1000L < System.currentTimeMillis()) {
            return JSONObject()
        }
        val now = System.currentTimeMillis()
        val pump = JSONObject()
        val battery = JSONObject()
        val status = JSONObject()
        val extended = JSONObject()
        try {
            battery.put("percent", pumpStatusData.batteryRemaining)
            status.put("status", pumpStatusData.pumpRunningState.status)
            extended.put("Version", version)
            try {
                extended.put("ActiveProfile", profileName)
            } catch (ignored: Exception) {
            }
            val tb = pumpSync.expectedPumpState().temporaryBasal
            if (tb != null) {
                extended.put("TempBasalAbsoluteRate", tb.convertedToAbsolute(now, profile))
                extended.put("TempBasalStart", dateUtil.dateAndTimeString(tb.timestamp))
                extended.put("TempBasalRemaining", tb.plannedRemainingMinutes)
            }
            val eb = pumpSync.expectedPumpState().extendedBolus
            if (eb != null) {
                extended.put("ExtendedBolusAbsoluteRate", eb.rate)
                extended.put("ExtendedBolusStart", dateUtil.dateAndTimeString(eb.timestamp))
                extended.put("ExtendedBolusRemaining", eb.plannedRemainingMinutes)
            }
            status.put("timestamp", dateUtil.toISOString(dateUtil.now()))
            pump.put("battery", battery)
            pump.put("status", status)
            pump.put("extended", extended)
            pump.put("reservoir", pumpStatusData.reservoirRemainingUnits)
            pump.put("clock", dateUtil.toISOString(dateUtil.now()))
        } catch (e: JSONException) {
            aapsLogger.error("Unhandled exception", e)
        }
        return pump
    }

    // FIXME i18n, null checks: iob, TDD
    override fun shortStatus(veryShort: Boolean): String {
        var ret = ""

        ret += if (pumpStatusData.lastConnection == 0L) {
            "LastConn: never\n"
        } else {
            val agoMin = ((System.currentTimeMillis() - pumpStatusData.lastConnection) / 60.0 / 1000.0).toInt()
            "LastConn: $agoMin min ago\n"
        }

        pumpStatusData.lastBolusTime?.let {
            if (it.time != 0L) {
                ret += "LastBolus: ${to2Decimal(pumpStatusData.lastBolusAmount!!)}U @${DateFormat.format("HH:mm", it)}\n"
            }
        }
        pumpSync.expectedPumpState().temporaryBasal?.let { ret += "Temp: ${it.toStringFull(dateUtil)}\n" }
        pumpSync.expectedPumpState().extendedBolus?.let { ret += "Extended: ${it.toStringFull(dateUtil)}\n" }
        ret += "IOB: ${pumpStatusData.iob}U\n"
        ret += "Reserv: ${to0Decimal(pumpStatusData.reservoirRemainingUnits)}U\n"
        ret += "Batt: ${pumpStatusData.batteryRemaining}\n"
        return ret
    }

    @Synchronized
    override fun deliverTreatment(detailedBolusInfo: DetailedBolusInfo): PumpEnactResult {
        return try {
            if (detailedBolusInfo.insulin == 0.0 && detailedBolusInfo.carbs == 0.0) {
                // neither carbs nor bolus requested
                aapsLogger.error("deliverTreatment: Invalid input")
                PumpEnactResult(injector).success(false).enacted(false).bolusDelivered(0.0).carbsDelivered(0.0)
                    .comment(R.string.invalid_input)
            } else if (detailedBolusInfo.insulin > 0) {
                // bolus needed, ask pump to deliver it
                deliverBolus(detailedBolusInfo)
            } else {
                detailedBolusInfo.timestamp = System.currentTimeMillis()

                // no bolus required, carb only treatment
                pumpSyncStorage.addCarbs(PumpDbEntryCarbs(detailedBolusInfo, this))

                val bolusingEvent = EventOverviewBolusProgress
                bolusingEvent.t = EventOverviewBolusProgress.Treatment(0.0, detailedBolusInfo.carbs.toInt(), detailedBolusInfo.bolusType === DetailedBolusInfo.BolusType.SMB, detailedBolusInfo.id)
                bolusingEvent.percent = 100
                rxBus.send(bolusingEvent)
                aapsLogger.debug(LTag.PUMP, "deliverTreatment: Carb only treatment.")
                PumpEnactResult(injector).success(true).enacted(true).bolusDelivered(0.0)
                    .carbsDelivered(detailedBolusInfo.carbs).comment(R.string.common_resultok)
            }
        } finally {
            triggerUIChange()
        }
    }

    protected fun refreshCustomActionsList() {
        rxBus.send(EventCustomActionsChanged())
    }

    override fun manufacturer(): ManufacturerType = pumpType.manufacturer!!
    override fun model(): PumpType = pumpType
    override fun canHandleDST(): Boolean = false

    protected abstract fun deliverBolus(detailedBolusInfo: DetailedBolusInfo): PumpEnactResult

    protected abstract fun triggerUIChange()

    private fun getOperationNotSupportedWithCustomText(resourceId: Int): PumpEnactResult =
        PumpEnactResult(injector).success(false).enacted(false).comment(resourceId)

    init {
        pumpDescription.fillFor(pumpType)
        this.pumpType = pumpType
    }
}
